File: ios/Sources/SIT/Services/LinkedInCSVParser.swift
import Foundation
import SwiftData

struct LinkedInCSVParser {

    /// Parses a LinkedIn Connections CSV export and returns Contact models.
    /// Expected headers: First Name, Last Name, Email Address, Company, Position
    ///
    /// The whole file is tokenized as a character stream (RFC 4180) rather than
    /// line-by-line, so a quoted field may contain embedded commas AND newlines
    /// without splitting one connection across two records. A leading UTF-8 BOM
    /// is stripped. Kept in lockstep with the Android LinkedInCSVParser.
    @discardableResult
    static func parse(url: URL, context: ModelContext) throws -> (imported: Int, skipped: Int) {
        let raw = try String(contentsOf: url, encoding: .utf8)
        let records = parseCSV(raw)

        guard let headerIndex = records.firstIndex(where: isHeaderRow) else {
            throw CSVError.noHeaderFound
        }

        let headers = records[headerIndex]
        let firstNameIdx = headers.firstIndex(of: "First Name") ?? 0
        let lastNameIdx  = headers.firstIndex(of: "Last Name") ?? 1
        let emailIdx     = headers.firstIndex(of: "Email Address")
        let companyIdx   = headers.firstIndex(of: "Company")
        let positionIdx  = headers.firstIndex(of: "Position")
        let phoneIdx     = headers.firstIndex(of: "Phone Number")

        var imported = 0
        var skipped = 0

        // Fetch existing fingerprints to detect duplicates. Mutated as we go so
        // duplicates *within* this same file are also caught, not just ones that
        // already existed in the store before the loop started.
        var seenFingerprints: Set<String> = {
            let descriptor = FetchDescriptor<Contact>()
            let all = (try? context.fetch(descriptor)) ?? []
            return Set(all.map(\.fingerprint).filter { !$0.isEmpty })
        }()

        for fields in records[(headerIndex + 1)...] {
            guard fields.contains(where: { !$0.isEmpty }) else { continue }
            guard fields.count > max(firstNameIdx, lastNameIdx) else { continue }

            let firstName    = fields[safe: firstNameIdx] ?? ""
            let lastName     = fields[safe: lastNameIdx] ?? ""
            if firstName.isEmpty && lastName.isEmpty { continue }

            let phoneNumbers = [fields[safe: phoneIdx] ?? ""].filter { !$0.isEmpty }
            let emails       = [fields[safe: emailIdx] ?? ""].filter { !$0.isEmpty }

            let fingerprint = ContactFingerprint.compute(
                firstName: firstName,
                lastName: lastName,
                phoneNumbers: phoneNumbers,
                emails: emails
            )

            // Skip if already exists (in the store or earlier in this file)
            if !fingerprint.isEmpty && seenFingerprints.contains(fingerprint) {
                skipped += 1
                continue
            }

            let contact = Contact(
                firstName:    firstName,
                lastName:     lastName,
                phoneNumbers: phoneNumbers,
                emails:       emails,
                company:      fields[safe: companyIdx] ?? "",
                jobTitle:     fields[safe: positionIdx] ?? "",
                importSource: .linkedin,
                fingerprint:  fingerprint
            )
            context.insert(contact)
            if !fingerprint.isEmpty { seenFingerprints.insert(fingerprint) }
            imported += 1
        }

        try context.save()
        print("Ticklr: Imported \(imported) contacts, skipped \(skipped) duplicates")
        return (imported: imported, skipped: skipped)
    }

    /// LinkedIn CSVs carry preamble/notes lines before the column row.
    private static func isHeaderRow(_ fields: [String]) -> Bool {
        fields.contains("First Name") && fields.contains("Last Name")
    }

    /// Tokenizes an entire CSV document into records of fields (RFC 4180).
    /// Quoted fields may contain commas, CR/LF, and escaped quotes ("").
    /// A leading UTF-8 BOM is stripped. Fields are whitespace-trimmed.
    private static func parseCSV(_ input: String) -> [[String]] {
        var text = Substring(input)
        if text.first == "\u{FEFF}" { text = text.dropFirst() }

        var records: [[String]] = []
        var record: [String] = []
        var current = ""
        var inQuotes = false

        func endField() {
            record.append(current.trimmingCharacters(in: .whitespacesAndNewlines))
            current = ""
        }
        func endRecord() {
            endField()
            records.append(record)
            record = []
        }

        var iterator = text.makeIterator()
        var pending: Character? = iterator.next()

        while let char = pending {
            pending = iterator.next()

            if inQuotes {
                if char == "\"" {
                    if pending == "\"" {           // escaped quote ("") → literal quote
                        current.append("\"")
                        pending = iterator.next()
                    } else {
                        inQuotes = false
                    }
                } else {
                    current.append(char)
                }
            } else {
                switch char {
                case "\"":
                    inQuotes = true
                case ",":
                    endField()
                case "\r":
                    endRecord()
                    if pending == "\n" { pending = iterator.next() }  // consume CRLF
                case "\n":
                    endRecord()
                default:
                    current.append(char)
                }
            }
        }
        // Flush the final field/record unless the file ended on a newline.
        if !current.isEmpty || !record.isEmpty {
            endRecord()
        }
        return records
    }

    enum CSVError: Error {
        case noHeaderFound
    }
}

private extension Array {
    subscript(safe index: Int?) -> Element? {
        guard let i = index, i >= 0, i < count else { return nil }
        return self[i]
    }
}

File: android/app/src/main/java/com/xaymaca/sit/service/LinkedInCSVParser.kt
package com.xaymaca.sit.service

import com.google.gson.Gson
import com.xaymaca.sit.data.model.Contact
import com.xaymaca.sit.data.model.ImportSource
import java.io.InputStream
import javax.inject.Inject
import javax.inject.Singleton
/**
 * Parses LinkedIn Connections CSV export.
 * Expected columns: First Name, Last Name, Email Address, Company, Position
 * LinkedIn CSV export path: linkedin.com → Me → Settings → Data privacy → Get a copy of your data
 *
 * The whole file is tokenised as a character stream (RFC 4180) rather than
 * line-by-line, so a quoted field may contain embedded commas AND newlines
 * without splitting one connection across two records. A leading UTF-8 BOM is
 * stripped so the first header column ("First Name") still matches.
 * Kept in lockstep with the iOS LinkedInCSVParser.
 */
@Singleton
class LinkedInCSVParser @Inject constructor() {

    private val gson = Gson()

    fun parse(inputStream: InputStream): List<Contact> {
        val text = inputStream.reader(Charsets.UTF_8).use { it.readText() }
        val records = parseCsv(text)

        // LinkedIn exports carry preamble/notes lines before the column row.
        val headerIdx = records.indexOfFirst { isHeaderRow(it) }
        if (headerIdx == -1) return emptyList()

        val headers = records[headerIdx]
        val firstNameIdx = headers.indexOfFirst { it.trim().equals("First Name", ignoreCase = true) }
        val lastNameIdx = headers.indexOfFirst { it.trim().equals("Last Name", ignoreCase = true) }
        val emailIdx = headers.indexOfFirst { it.trim().equals("Email Address", ignoreCase = true) }
        val companyIdx = headers.indexOfFirst { it.trim().equals("Company", ignoreCase = true) }
        val positionIdx = headers.indexOfFirst { it.trim().equals("Position", ignoreCase = true) }
        val phoneIdx = headers.indexOfFirst { it.trim().equals("Phone Number", ignoreCase = true) }

        if (firstNameIdx == -1 && lastNameIdx == -1) return emptyList()

        val result = mutableListOf<Contact>()
        for (cols in records.subList(headerIdx + 1, records.size)) {
            if (cols.all { it.isBlank() }) continue

            fun col(idx: Int): String = if (idx >= 0 && idx < cols.size) cols[idx].trim() else ""

            val firstName = col(firstNameIdx)
            val lastName = col(lastNameIdx)
            if (firstName.isBlank() && lastName.isBlank()) continue

            val email = col(emailIdx)
            val emails: List<String> = if (email.isNotBlank()) listOf(email) else emptyList()

            val phone = col(phoneIdx)
            val phones: List<String> = if (phone.isNotBlank()) listOf(phone) else emptyList()

            val phoneNumbersJson = gson.toJson(phones)
            val emailsJson = gson.toJson(emails)
            val fingerprint = ContactFingerprint.compute(firstName, lastName, phoneNumbersJson, emailsJson)

            result.add(
                Contact(
                    firstName = firstName,
                    lastName = lastName,
                    phoneNumbers = phoneNumbersJson,
                    emails = emailsJson,
                    company = col(companyIdx),
                    jobTitle = col(positionIdx),
                    importSource = ImportSource.LINKEDIN.name,
                    createdAt = System.currentTimeMillis(),
                    fingerprint = fingerprint
                )
            )
        }

        return result
    }

    private fun isHeaderRow(fields: List<String>): Boolean =
        fields.any { it.trim().equals("First Name", ignoreCase = true) } &&
            fields.any { it.trim().equals("Last Name", ignoreCase = true) }

    /**
     * Tokenises an entire CSV document into records of fields (RFC 4180).
     * Quoted fields may contain commas, CR/LF, and escaped quotes ("").
     * A leading UTF-8 BOM is stripped. Blank trailing lines are dropped.
     */
    private fun parseCsv(input: String): List<List<String>> {
        val text = input.removePrefix("\uFEFF")
        val records = mutableListOf<List<String>>()
        var record = mutableListOf<String>()
        val current = StringBuilder()
        var inQuotes = false
        var i = 0

        fun endField() {
            record.add(current.toString())
            current.clear()
        }
        fun endRecord() {
            endField()
            records.add(record)
            record = mutableListOf()
        }

        while (i < text.length) {
            val ch = text[i]
            when {
                inQuotes -> when {
                    ch == '"' && i + 1 < text.length && text[i + 1] == '"' -> {
                        current.append('"'); i += 2
                    }
                    ch == '"' -> { inQuotes = false; i++ }
                    else -> { current.append(ch); i++ }
                }
                ch == '"' -> { inQuotes = true; i++ }
                ch == ',' -> { endField(); i++ }
                ch == '\r' -> {
                    endRecord()
                    i += if (i + 1 < text.length && text[i + 1] == '\n') 2 else 1
                }
                ch == '\n' -> { endRecord(); i++ }
                else -> { current.append(ch); i++ }
            }
        }
        // Flush the final field/record unless the file ended on a newline.
        if (current.isNotEmpty() || record.isNotEmpty()) {
            endRecord()
        }
        return records
    }
}
